File: ios/Noa/Noa/Frame Assets/Scripts/audio.lua
require("states")

function StartRecording(state, gfx, send_message)
    if state:on_entry() then
        frame.microphone.record(0.05, 8000, 8)
        if state.previous_state == State.SendImage then
            send_message("ien:") -- continue image-and-prompt flow
        else
            send_message("ast:") -- *prompt only* (erases image data)
        end
        gfx:clear_response()
        gfx:set_prompt("Listening [     ]")
    end
    state:after(1, State.SendAudio)
end

function SendAudio(state, gfx, send_message)
    if state:has_been() > 5 then
        gfx:set_prompt("Waiting for openAI")
    elseif state:has_been() > 4 then
        gfx:set_prompt("Listening [=====]")
    elseif state:has_been() > 3 then
        gfx:set_prompt("Listening [==== ]")
    elseif state:has_been() > 2 then
        gfx:set_prompt("Listening [===  ]")
    elseif state:has_been() > 1 then
        gfx:set_prompt("Listening [==   ]")
    else
        gfx:set_prompt("Listening [=    ]")
    end

    local samples = (frame.bluetooth.max_length() - 4) // 10
    local chunk1 = frame.microphone.read(samples)
    local chunk2 = frame.microphone.read(samples)

    if chunk1 == nil then
        send_message("aen:")
        state:after(0, State.WaitForPing)
    elseif chunk2 == nil then
        local dt = "dat:" .. table.concat(chunk1)
        print(#dt)
        send_message(dt)
    else
        local dt = "dat:" .. table.concat(chunk1) .. table.concat(chunk2)
        print(#dt)
        send_message(dt)
    end
end

File: ios/Noa/Noa/Monocle Assets/Scripts/audio.py
import bluetooth
import microphone
import states

def start_recording(state, gfx, send_message):
    if state.on_entry():
        microphone.record(seconds=6.0, bit_depth=8, sample_rate=8000)
        if state.previous_state == state.SendImage:
            send_message(b"ien:") # continue image-and-prompt flow
        else:
            send_message(b"ast:") # *prompt only* (erases image data)
        gfx.clear_response()
        gfx.set_prompt("Listening [     ]")
    state.after(1000, state.SendAudio)

def send_audio(state, gfx, send_message):
    if state.has_been() > 5000:
        gfx.set_prompt("Waiting for openAI")
    elif state.has_been() > 4000:
        gfx.set_prompt("Listening [=====]")
    elif state.has_been() > 3000:
        gfx.set_prompt("Listening [==== ]")
    elif state.has_been() > 2000:
        gfx.set_prompt("Listening [===  ]")
    elif state.has_been() > 1000:
        gfx.set_prompt("Listening [==   ]")
    else:
        gfx.set_prompt("Listening [=    ]")

    samples = (bluetooth.max_length() - 4) // 2
    chunk1 = microphone.read(samples)
    chunk2 = microphone.read(samples)

    if chunk1 == None:
        send_message(b"aen:")
        state.after(0, state.WaitForPing)
    elif chunk2 == None:
        send_message(b"dat:" + chunk1)
    else:
        send_message(b"dat:" + chunk1 + chunk2)
